### PR TITLE
FEATURE: add new custom status filters

### PR DESCRIPTION
**Pull request by ZogStriP**
Those can be used in /filter route.

Depends on https://github.com/discourse/discourse/pull/26770

Internal ref. /t/127278

### CODE REVIEW

File: plugin.rb [hunk 1]
```diff
@@ -310,15 +310,16 @@ def self.skip_db?
   end
 
   if respond_to?(:register_modifier)
-    register_modifier(:search_rank_sort_priorities) do |priorities, search|
+    register_modifier(:search_rank_sort_priorities) do |priorities, _search|
```
Comment (ZogStriP) [inline plugin.rb:313]: Added the "`_`" prefix to signal that this argument isn't being used.

File: plugin.rb [hunk 2]
```diff
@@ -310,15 +310,16 @@ def self.skip_db?
...
       if SiteSetting.prioritize_solved_topics_in_search
         condition = <<~SQL
-          EXISTS
-            (
-              SELECT 1 FROM topic_custom_fields f
-              WHERE topics.id = f.topic_id
-              AND f.name = '#{::DiscourseSolved::ACCEPTED_ANSWER_POST_ID_CUSTOM_FIELD}'
-            )
+          EXISTS (
```
Comment (ZogStriP) [inline plugin.rb:316]: Formatted the query and added the "`AND value IS NOT NULL`" condition that is being used everywhere else

File: plugin.rb [hunk 3]
```diff
@@ -344,82 +345,72 @@ def self.skip_db?
...
-  if Search.respond_to? :advanced_filter
-    Search.advanced_filter(/status:solved/) do |posts|
-      posts.where(
-        "topics.id IN (
-        SELECT tc.topic_id
-        FROM topic_custom_fields tc
-        WHERE tc.name = '#{::DiscourseSolved::ACCEPTED_ANSWER_POST_ID_CUSTOM_FIELD}' AND
-                        tc.value IS NOT NULL
-        )",
+  solved_callback = ->(scope) do
```
Comment (ZogStriP) [inline plugin.rb:348]: We have several ways to filter by solved/unsolved status (in search, in filter, in topic lists, etc...).

I've extracted both solved and unsolved "queries" so they can be used everywhere.

File: plugin.rb [hunk 4]
```diff
@@ -344,82 +345,72 @@ def self.skip_db?
...
         )
-      end
+      SQL
     end
+
+    scope
   end
 
-  if Discourse.has_needed_version?(Discourse::VERSION::STRING, "1.8.0.beta6")
+  if respond_to? :register_custom_filter_by_status
```
Comment (ZogStriP) [inline plugin.rb:395]: Using the new API introduced in https://github.com/discourse/discourse/pull/26770

File: plugin.rb [hunk 5]
```diff
@@ -344,82 +345,72 @@ def self.skip_db?
...
+    register_custom_filter_by_status("solved", &solved_callback)
+    register_custom_filter_by_status("unsolved", &unsolved_callback)
+  end
+
+  if respond_to? :register_search_advanced_filter
```
Comment (ZogStriP) [inline plugin.rb:400]: Used the "plugin" API instead of reaching directly for the `Search` class.

File: plugin.rb [hunk 6]
```diff
@@ -432,19 +423,27 @@ def self.skip_db?
   if Search.respond_to? :preloaded_topic_custom_fields
     Search.preloaded_topic_custom_fields << ::DiscourseSolved::ACCEPTED_ANSWER_POST_ID_CUSTOM_FIELD
   end
-
-  if CategoryList.respond_to?(:preloaded_topic_custom_fields)
+  if CategoryList.respond_to? :preloaded_topic_custom_fields
```
Comment (ZogStriP) [inline plugin.rb:426]: Removed parenthesis to match surrounding styling.

File: plugin.rb [hunk 7]
```diff
@@ -432,19 +423,27 @@ def self.skip_db?
...
 
-  on(:filter_auto_bump_topics) { |_category, filters| filters.push(->(r) { r.where(<<~SQL) }) }
-        NOT EXISTS(
-          SELECT 1 FROM topic_custom_fields
-          WHERE topic_id = topics.id
-          AND name = '#{::DiscourseSolved::ACCEPTED_ANSWER_POST_ID_CUSTOM_FIELD}'
-          AND value IS NOT NULL
-        )
-      SQL
+  on(:filter_auto_bump_topics) do |_category, filters|
```
Comment (ZogStriP) [inline plugin.rb:430]: Tweaked the formatting a little bit so it's less confusing 🤷

**Comment by jjaffeux** [comment]
How about using discourse_compatibility so we dont have to do all the ifs everywhere?

**Comment by ZogStriP** [comment]
@jjaffeux Was https://github.com/discourse/discourse-solved/pull/286/commits/2dbb02f3fe91f7e9145a396de27add69ccb7cc73 what you had in mind? 🤔

**Comment by jjaffeux** [comment]
> @jjaffeux Is [2dbb02f](https://github.com/discourse/discourse-solved/commit/2dbb02f3fe91f7e9145a396de27add69ccb7cc73) what you had in mind? 🤔

Yes we know these APIs are in discourse at this version, so it should be safe

**Review by jjaffeux — APPROVED** [review]

**Pull request merged**